### PR TITLE
Resolve #872

### LINKED ISSUE #872: rand(Categorical) is not safe for Float32
When I do `rand(::Categorical)`, I meet error saying that `cp += p[i +=1]` is visiting an index out of the range of `p` (when `eltype(p)` is `Float32`), in this function https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/discrete/discretenonparametric.jl#L79-L89

```julia
function rand(rng::AbstractRNG, d::DiscreteNonParametric{T,P}) where {T,P}
    x = support(d)
    p = probs(d)
    draw = rand(rng, P)
    cp = zero(P)
    i = 0
    while cp < draw
        cp += p[i +=1]
    end
    x[i]
end
```

I'm guessing when `eltype(p)` is `Float32`, `cp < draw` has chance of being never `true` thus give this bound error. I tried to convert `p` to `Vector{Flaot64}` and re-normalise, and I didn't see this error after doing so.

### PR DESCRIPTION
**Pull request by xukai92**
Apply the trick from @mschauer (https://github.com/JuliaStats/Distributions.jl/issues/872#issuecomment-486580512) to resolve #872.

### CODE REVIEW

**Comment by matbesancon** [comment]
@xukai92 cool thanks, could you add the test that was failing before in the tests?

**Comment by matbesancon** [comment]
ping @xukai92

**Comment by xukai92** [comment]
Woud the example below do the job?

```julia
p = [1 - eps(Float32), eps(Float32)]
d = Categorical(p)
v = [rand(d) for _ = 1:100_000_000]
```

Which with the current master throws

```
ERROR: BoundsError: attempt to access 2-element Base.OneTo{Int64} at index [0]
Stacktrace:
 [1] throw_boundserror(::Base.OneTo{Int64}, ::Int64) at ./abstractarray.jl:484
 [2] getindex at ./range.jl:622 [inlined]
 [3] rand(::Random.MersenneTwister, ::DiscreteNonParametric{Int64,Float32,Base.OneTo{Int64},Array{Float32,1}}) at /Users/kai/.julia/packages/Distributions/0Wogo/src/univariate/discrete/discretenonparametric.jl:88
 [4] rand(::DiscreteNonParametric{Int64,Float32,Base.OneTo{Int64},Array{Float32,1}}) at /Users/kai/.julia/packages/Distributions/0Wogo/src/univariate/discrete/discretenonparametric.jl:91
 [5] (::getfield(Main, Symbol("##13#14")))(::Int64) at ./none:0
 [6] iterate at ./generator.jl:47 [inlined]
 [7] collect_to!(::Array{Int64,1}, ::Base.Generator{UnitRange{Int64},getfield(Main, Symbol("##13#14"))}, ::Int64, ::Int64) at ./array.jl:651
 [8] collect_to_with_first!(::Array{Int64,1}, ::Int64, ::Base.Generator{UnitRange{Int64},getfield(Main, Symbol("##13#14"))}, ::Int64) at ./array.jl:630
 [9] collect(::Base.Generator{UnitRange{Int64},getfield(Main, Symbol("##13#14"))}) at ./array.jl:611
 [10] top-level scope at none:0
```
and with current PR works fine.

**Comment by matbesancon** [comment]
Yep great

File: test/discretenonparametric.jl [hunk 1]
```diff
@@ -107,4 +107,9 @@ d3 = fit(DiscreteNonParametric, xs)
 @test support(d3) == support(d1)
 @test probs(d3) ≈ probs(d1)
 
+# Numerical stability; see issue #872 and PR #926
+p = [1 - eps(Float32), eps(Float32)]
+d = Categorical(p)
+@test ([rand(d) for _ = 1:100_000_000]; true)
```
Comment (mschauer) [inline test/discretenonparametric.jl]:
```suggestion
@test ([rand(d) for _ = 1:100_000]; true)   
```
Lets save electricity

Comment (xukai92) [reply test/discretenonparametric.jl]: Great!

**Comment by matbesancon** [comment]
merging in 12 hours if no objection

**Pull request merged**